tiledb/sm/serialization/array.cc

```diff
@@ -0,0 +1,460 @@
+  }
+  array_builder->setStartTimestamp(array->timestamp_start());
+  array_builder->setEndTimestamp(array->timestamp_end());
+
+  ArraySchema* array_schema_latest = array->array_schema_latest();
+  auto array_schema_latest_builder = array_builder->initArraySchemaLatest();
+  RETURN_NOT_OK(array_schema_to_capnp(
+      array_schema_latest, &array_schema_latest_builder, client_side));
+
+  const auto& array_schemas_all = array->array_schemas_all();
+  auto array_schemas_all_builder = array_builder->initArraySchemasAll();
+  auto entries_builder =
+      array_schemas_all_builder.initEntries(array_schemas_all.size());
+  uint64_t i = 0;
+  for (const auto& schema : array_schemas_all) {
+    auto entry = entries_builder[i++];
+    entry.setKey(schema.first);
+    auto schema_builder = entry.initValue();
+    RETURN_NOT_OK(array_schema_to_capnp(
+        schema.second.get(), &schema_builder, client_side));
+  }
+
+  auto nonempty_domain_builder = array_builder->initNonEmptyDomain();
+  RETURN_NOT_OK(
+      utils::serialize_non_empty_domain(nonempty_domain_builder, array));
+
+  auto array_metadata_builder = array_builder->initArrayMetadata();
+  RETURN_NOT_OK(array_metadata_to_capnp(array, &array_metadata_builder));
+
+  return Status::Ok();
+}
+
+Status array_from_capnp(
+    const capnp::Array::Reader& array_reader, Array* array) {
+  // The serialized URI is set if it exists
+  // this is used for backwards compatibility with pre TileDB 2.5 clients that
+  // want to serialized a query object TileDB >= 2.5 no longer needs to receive
+  // the array URI
+  if (array_reader.hasUri()) {
+    RETURN_NOT_OK(array->set_uri_serialized(array_reader.getUri().cStr()));
+  }
+  RETURN_NOT_OK(array->set_timestamp_start(array_reader.getStartTimestamp()));
+  RETURN_NOT_OK(array->set_timestamp_end(array_reader.getEndTimestamp()));
+
+  if (array_reader.hasArraySchemasAll()) {
+    std::unordered_map<std::string, tdb_shared_ptr<ArraySchema>> all_schemas;
+    auto all_schemas_reader = array_reader.getArraySchemasAll();
+
+    if (all_schemas_reader.hasEntries()) {
+      auto entries = array_reader.getArraySchemasAll().getEntries();
+      for (auto array_schema_build : entries) {
+        tdb_unique_ptr<ArraySchema> schema;
+        RETURN_NOT_OK(
+            array_schema_from_capnp(array_schema_build.getValue(), &schema));
+        schema->set_array_uri(array->array_uri());
+        all_schemas[array_schema_build.getKey()] = std::move(schema);
+      }
+    }
+    RETURN_NOT_OK(array->set_array_schemas_all(all_schemas));
+  }
+
+  if (array_reader.hasArraySchemaLatest()) {
+    tdb_unique_ptr<ArraySchema> array_schema_latest;
+    auto array_schema_latest_reader = array_reader.getArraySchemaLatest();
+    RETURN_NOT_OK(array_schema_from_capnp(
+        array_schema_latest_reader, &array_schema_latest));
+    array_schema_latest->set_array_uri(array->array_uri());
+    RETURN_NOT_OK(
+        array->set_array_schema_latest(array_schema_latest.release()));
```

Comment:
I'm not 100% sure of how this works, but I do believe that for remote arrays, the storage manager doesn't own the latest array schema, so in the set_array_schema_latest function, you might want to delete the old schema before replacing the pointer?